File: LeetCode/C/146-design-lru缓存机制.cpp
/*
 * @lc app=leetcode.cn id=146 lang=cpp
 *
 * [146] LRU缓存机制
 */
#include "main.h"

// @lc code=start
class LRUCache {
  public:
    LRUCache(int capacity) {
        this->size = 0;
        this->capacity = capacity;
        this->head = new Node(-1, -1);
        this->tail = new Node(-1, -1);
        head->next = tail;
        tail->prev = head;
    }

    int get(int key) {
        if (cache.count(key)) {
            moveToHead(cache[key]);
            return cache[key]->value;
        } else
            return -1;
    }

    void put(int key, int value) {
        if (cache.count(key)) {
            cache[key]->value = value;
            moveToHead(cache[key]);
        } else {
            auto newNode = new Node(key, value);
            cache[key] = newNode;
            addNode(newNode);
            ++size;
            if (size > capacity) {
                int k = tail->prev->key;
                removeNode(cache[k]);
                cache.erase(cache.find(k));
                --size;
            }
        }
    }

  private:
    struct Node {
        int key;
        int value;
        Node *prev, *next;
        Node(int k, int v) {
            key = k;
            value = v;
        }
    };

    Node *head, *tail;
    int size, capacity;

    unordered_map<int, Node *> cache;

    void addNode(Node *node) {
        node->prev = head;
        node->next = head->next;
        head->next->prev = node;
        head->next = node;
    }

    void removeNode(Node *node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
        delete node;
    }

    void moveToHead(Node *node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
        addNode(node);
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end

File: LeetCode/Go/146-design-lru缓存机制.go
/*
 * @lc app=leetcode.cn id=146 lang=golang
 *
 * [146] LRU缓存机制
 */
package leetcode

// @lc code=start
type DNode struct {
	key   int
	value int
	prev  *DNode
	next  *DNode
}

type LRUCache struct {
	cache map[int]*DNode

	head *DNode
	tail *DNode

	size     int
	capacity int
}

func Constructor(capacity int) LRUCache {
	lru := LRUCache{
		size:     0,
		capacity: capacity,
		head:     &DNode{key: -1, value: -1},
		tail:     &DNode{key: -1, value: -1},
		cache:    make(map[int]*DNode),
	}
	lru.head.next = lru.tail
	lru.tail.prev = lru.head
	return lru
}

func (lru *LRUCache) addNode(node *DNode) {
	node.prev = lru.head
	node.next = lru.head.next
	lru.head.next.prev = node
	lru.head.next = node
}

func (lru *LRUCache) removeNode(node *DNode) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (lru *LRUCache) moveToHead(node *DNode) {
	node.prev.next = node.next
	node.next.prev = node.prev
	lru.addNode(node)
}

func (this *LRUCache) Get(key int) int {
	if _, ok := this.cache[key]; ok {
		this.moveToHead(this.cache[key])
		return this.cache[key].value
	} else {
		return -1
	}
}

func (this *LRUCache) Put(key int, value int) {
	if _, ok := this.cache[key]; ok {
		this.cache[key].value = value
		this.moveToHead(this.cache[key])
	} else {
		node := &DNode{key: key, value: value}
		this.cache[key] = node
		this.addNode(node)
		this.size++
		if this.size > this.capacity {
			k := this.tail.prev.key
			this.removeNode(this.cache[k])
			delete(this.cache, k)
			this.size--
		}
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
// @lc code=end
